core/src/main/java/org/axonframework/messaging/annotation/AnnotatedHandlerInspector.java

```diff
@@ -68,64 +71,85 @@ private AnnotatedHandlerInspector(Class<T> inspectedType,
+                                                              registry));
         }
         //noinspection unchecked
         return registry.get(inspectedType);
     }
 
     private static <T> AnnotatedHandlerInspector<T> initialize(Class<T> inspectedType,
                                                                ParameterResolverFactory parameterResolverFactory,
+                                                               HandlerDefinition handlerDefinition,
                                                                Map<Class<?>, AnnotatedHandlerInspector> registry) {
         List<AnnotatedHandlerInspector<? super T>> parents = new ArrayList<>();
         for (Class<?> iFace : inspectedType.getInterfaces()) {
             //noinspection unchecked
-            parents.add(createInspector(iFace, parameterResolverFactory, registry));
+            parents.add(createInspector(iFace,
+                                        parameterResolverFactory,
+                                        handlerDefinition,
+                                        registry));
         }
         if (inspectedType.getSuperclass() != null && !Object.class.equals(inspectedType.getSuperclass())) {
-            parents.add(createInspector(inspectedType.getSuperclass(), parameterResolverFactory, registry));
+            parents.add(createInspector(inspectedType.getSuperclass(),
+                                        parameterResolverFactory,
+                                        handlerDefinition,
+                                        registry));
         }
-        AnnotatedHandlerInspector<T> inspector =
-                new AnnotatedHandlerInspector<>(inspectedType, parents, parameterResolverFactory, registry);
-        inspector.initializeMessageHandlers(parameterResolverFactory);
+        AnnotatedHandlerInspector<T> inspector = new AnnotatedHandlerInspector<>(inspectedType,
+                                                                                 parents,
+                                                                                 parameterResolverFactory,
+                                                                                 handlerDefinition,
+                                                                                 registry);
+        inspector.initializeMessageHandlers(parameterResolverFactory, handlerDefinition);
         return inspector;
     }
 
-    private void initializeMessageHandlers(ParameterResolverFactory parameterResolverFactory) {
-        List<HandlerDefinition> definitions = new ArrayList<>();
-        ServiceLoader.load(HandlerDefinition.class).forEach(definitions::add);
-        List<HandlerEnhancerDefinition> wrapperDefinitions = new ArrayList<>();
-        ServiceLoader.load(HandlerEnhancerDefinition.class).forEach(wrapperDefinitions::add);
+    private void initializeMessageHandlers(ParameterResolverFactory parameterResolverFactory,
```

Comment:
This function is the only reason why you need to pass along the `HandlerDefinition` everywhere, right? ^_^;

Reply:
Correct! Are you just stating, or you have some better approach? 😄

Reply:
This is me tracing my own mind through the review and trying to understand why so much has changed whilst the added functionality is quite minimal ^_^;

So, stating, don't know a better solution :o